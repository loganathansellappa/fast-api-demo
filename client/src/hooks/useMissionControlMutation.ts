import { useAxiosClient } from '../context/AxiosClientContext.ts';
import { useMutation, useQueryClient } from 'react-query';
import { missionControlKeys } from '../services/queryKeyFactory.ts';
import { MissionState, MutationType } from '../utils/HelperUtils.ts';

export type MissionControlMutation = {
  missionState?: MissionState;
  mutationType: MutationType;
  url: string;
  description?: string;
  title?: string;
};

export type MissionControlInput = {
  mutationType: MutationType;
  id?: number;
  missionState?: MissionState;
  mission_state?: MissionState;
  description?: string;
  title?: string;
};

type RequestData = {
  url: string;
  method: MutationType;
  mission_state?: MissionState;
  data?: Record<string, unknown>;
};

export function useMissionControlUpdate() {
  const axiosClientService = useAxiosClient();
  const queryClient = useQueryClient();

  const mutation = useMutation({
    mutationFn: (requestData: MissionControlMutation) => {
      const { url, mutationType, missionState, ...data } = requestData;
      const requestObject: RequestData = {
          url: url,
          method: mutationType
      }
      if (mutationType !== MutationType.DELETE) {
        requestObject.data  = {
          ...data,
          mission_state: missionState
        }
      }
      return axiosClientService.axiosClient.request({...requestObject })
    },
    onSuccess: () => {
      queryClient.cancelQueries({ queryKey: missionControlKeys.all() });
      queryClient.invalidateQueries({ queryKey: missionControlKeys.all() });
    },
    onError: (error) => {
      console.error(error);
    }
  });

  const mutateMissionControl = ({ id, missionState = MissionState.PreFlight, mutationType, description, title }: MissionControlInput) => {
    const url = mutationType === MutationType.POST ? `flight_missions` : `flight_missions/${id}`;
    if (mutationType === MutationType.DELETE) {
      mutation.mutate({ mutationType, url });
    }  else {
      mutation.mutate({ missionState, mutationType, description, title, url });
    }

  };

  return {
    mutate: mutateMissionControl,
    isLoading: mutation.isLoading,
    isError: mutation.isError,
    error: mutation.error,
    isSuccess: mutation.isSuccess,
  };
}
