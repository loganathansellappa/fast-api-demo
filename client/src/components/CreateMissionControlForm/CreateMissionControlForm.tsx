import { MissionControlInput, useMissionControlUpdate } from '../../hooks/useMissionControlMutation.ts';
import { MissionState, MutationType, showErrorToast, showSuccessToast } from '../../utils/HelperUtils.ts';
import { useForm } from 'react-hook-form';
import { useId } from 'react';

interface FormInputs {
  title: string
  description: string
}
interface CreateMissionControlFormProps {
  onCancel: () => void;
}
export const CreateMissionControlForm = ({ onCancel }: CreateMissionControlFormProps ) => {
  const {  mutate, isError, isSuccess, error } = useMissionControlUpdate();
  const uId = useId();

  const { register, formState: { errors }, handleSubmit } = useForm<FormInputs>();

  if(isError) {
    showErrorToast(`Mission Creation Failed ${error}`, `createMissionControlFailed${uId}`);
  }

  if(isSuccess) {
    showSuccessToast('Mission Creation Success', `createMissionControlSuccess${uId}`);
    onCancel();
  }
  const onSubmit = (data: FormInputs) => {
    const requestData: MissionControlInput = {
      title: data.title,
      description: data.description,
      missionState: MissionState.PreFlight,
      mutationType: MutationType.POST,
    }
    mutate(requestData);
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
        <div className="mb-4">
          <label htmlFor="title" className="block text-gray-700 font-semibold mb-2">Title:</label>
          <input
            {...register('title', { required: 'This is required.' })}
            type="text"
            id="title"
            className="w-full border border-gray-300 rounded-md px-3 py-2 focus:outline-none focus:border-blue-500"
          />
          <p className="text-red-500">{errors?.title?.message}</p>
        </div>
        <div className="mb-4">
          <label htmlFor="description" className="block text-gray-700 font-semibold mb-2">Description:</label>
          <textarea
            {...register('description', { required: 'This is required.' })}
            className="w-full border border-gray-300 rounded-md px-3 py-2 focus:outline-none focus:border-blue-500"
            id="description"
          ></textarea>
          <p className="text-red-500">{errors?.description?.message}</p>

        </div>
        <div className="flex justify-end">
          <button
            type="button"
            onClick={onCancel}
            className="mr-2 px-4 py-2 bg-gray-300 hover:bg-gray-400 text-gray-800 font-semibold rounded-md focus:outline-none focus:bg-gray-400"
          >
            Cancel
          </button>
          <button
            type="submit"
            className="px-4 py-2 bg-blue-800 hover:bg-blue-600 text-white font-semibold rounded-md focus:outline-none focus:bg-blue-600"
          >
            Create
          </button>
        </div>
      </form>
  );
};
