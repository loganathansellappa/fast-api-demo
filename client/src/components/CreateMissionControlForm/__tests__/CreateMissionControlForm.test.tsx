import { render } from '@testing-library/react';
import { CreateMissionControlForm, CreateMissionControlFormProps } from '../CreateMissionControlForm.tsx';
import * as Hooks from '../../../hooks/useMissionControlMutation.ts';

describe("CreateMissionControlForm", () => {
  let element: { container: HTMLElement };
  const clickHandler = jest.fn();
  const submitHandler = jest.fn();

  beforeEach(() => {
    jest
      .spyOn(Hooks, "useMissionControlUpdate")
      .mockReturnValue(
        {
          mutate: submitHandler,
          isLoading: false,
          isError: false,
          error: false,
          isSuccess: false,
        }
      );
    const props: CreateMissionControlFormProps = {
      onCancel: clickHandler,
    };
    element = render(<CreateMissionControlForm {...props} />);
  });

  test("renders CreateMissionControlForm component", () => {
    expect(element.container.firstChild).toMatchSnapshot();
  });

  test("displays form", () => {
    expect(
      element.container.getElementsByClassName("mission-title")!.length,
    ).toBe(1);
    expect(
      element.container.getElementsByClassName("mission-description")!.length,
    ).toBe(1);
  });

  test("trigger click on cancel button click", () => {
    const cancelButton = element.container.getElementsByClassName("cancel-create")[0] as HTMLElement;
    cancelButton.click();
    expect(clickHandler).toBeCalledTimes(1);
  });
});