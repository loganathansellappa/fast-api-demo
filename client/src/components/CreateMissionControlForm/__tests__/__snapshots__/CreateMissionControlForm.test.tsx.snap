// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateMissionControlForm renders CreateMissionControlForm component 1`] = `
<form>
  <div
    class="mb-4"
  >
    <label
      class="block text-gray-700 font-semibold mb-2"
      for="title"
    >
      Title:
    </label>
    <input
      class="mission-title w-full border border-gray-300 rounded-md px-3 py-2 focus:outline-none focus:border-blue-500"
      id="title"
      name="title"
      type="text"
    />
    <p
      class="text-red-500"
    />
  </div>
  <div
    class="mb-4"
  >
    <label
      class="block text-gray-700 font-semibold mb-2"
      for="description"
    >
      Description:
    </label>
    <textarea
      class="mission-description w-full border border-gray-300 rounded-md px-3 py-2 focus:outline-none focus:border-blue-500"
      id="description"
      name="description"
    />
    <p
      class="text-red-500"
    />
  </div>
  <div
    class="flex justify-end"
  >
    <button
      class="cancel-create mr-2 px-4 py-2 bg-gray-300 hover:bg-gray-400 text-gray-800 font-semibold rounded-md focus:outline-none focus:bg-gray-400"
      type="button"
    >
      Cancel
    </button>
    <button
      class="submit-create-mission px-4 py-2 bg-blue-800 hover:bg-blue-600 text-white font-semibold rounded-md focus:outline-none focus:bg-blue-600"
      type="submit"
    >
      Create
    </button>
  </div>
</form>
`;
